port/zos390/omrvmem.c

```diff
@@ -1296,24 +1297,31 @@ omrvmem_default_large_page_size_ex(struct OMRPortLibrary *portLibrary, uintptr_t
 	}
 #endif /* OMR_ENV_DATA64 */
 
-	while (0 != PPG_vmem_pageSize[index]) {
-		if (0 != (OMRPORT_VMEM_MEMORY_MODE_EXECUTE & mode)) {
-			if ((ONE_M == PPG_vmem_pageSize[index]) &&
-				(0 != (OMRPORT_VMEM_PAGE_FLAG_PAGEABLE & PPG_vmem_pageFlags[index]))
-			) {
-				break;
-			}
+	/* Determine maximum pageable, and fixed large pages available. */
+	for (int i = 0; (0 != PG_vmem_pageSize[i]) && (ONE_M >= PPG_vmem_pageSize[index]), i++) {
+
+		if (0 != (OMRPORT_VMEM_PAGE_FLAG_PAGEABLE & PPG_vmem_pageFlags[i])) {
+			/* Pageable. Keep track of largest pageable size */
+			maxPageableIndex = i;
+		}
 #if defined(OMR_ENV_DATA64)
-		} else {
-			if ((ONE_M == PPG_vmem_pageSize[index]) &&
-				(0 != (OMRPORT_VMEM_PAGE_FLAG_FIXED & PPG_vmem_pageFlags[index]))
-			) {
-				break;
-			}
-#endif
+		else if (0 != (OMRPORT_VMEM_PAGE_FLAG_FIXED & PPG_vmem_pageFlags[i])) {
+			/* Fixed. Keep track of largest fixed size */
+			index = i;
 		}
-		index++;
+#endif /* OMR_ENV_DATA64 */
+	}
+
+	/* 
+	 * When mode is OMRPORT_VMEM_MEMORY_EXECUTE we will only consider pagable large pages.
+	 * When flag is OMRPORT_VMEM_PAGE_FLAG_PAGEABLE_PREFERABLE, we prefer pageable unless fixed is larger.
+	 * NOTE: 32 bit only supports Pageable large pages. OMRPORT_VMEM_MEMORY_MODE_EXECUTE is guaranteed to be high during 32bit execution.
+	 */
+	if ((0 != (OMRPORT_VMEM_MEMORY_MODE_EXECUTE & mode)) ||
+	   ((OMRPORT_VMEM_PAGE_FLAG_PAGEABLE_PREFERABLE == *pageFlags) && (PPG_vmem_pageSize[index] >= PPG_vmem_pageSize[maxFixedIndex]))) {
+		index = maxPageableIndex;
 	}
```

Comment:
Is `pageFlags` a bitvector of flags? Just checking whether we should be checking the `OMRPORT_VMEM_PAGE_FLAG_PAGEABLE_PREFERABLE` bit here rather than equality testing.

Reply:
It is a bitvector, where each bit represents a page type. We can definitely just check the bit. The same can be said for all other places in the port lib where we compare page flags.

Reply:
Seems I also forgot to check for a null pageFlags here. Added that check. 

See: https://github.com/eclipse/omr/pull/4762/commits/bbf1de405637508ca24b4ac4e71eb22cf6e81738